sys/src/libc/port/reallocarray.c

```diff
@@ -28,6 +28,17 @@
 void *
 reallocarray(void *optr, size_t nmemb, size_t size)
 {
+	/* if both size or nmemb are 0, and realloc is called with
+	 * 0, it's equivalent to freeing the memory. realloc is even
+	 * allowed to return nil. Hence this function needs to distinguish
+	 * between 'pointer to 0 length array' and 'nil pointer'. In the
+	 * event that nmemb * size is zero, we resize to 1 byte, which
+	 * is usable as a pointer to a 0-length array, and further usable
+	 * in later realloc calls.
+	 */
+	if (nmemb * size == 0)
+		return realloc(optr, 1);
+
 	if ((nmemb >= MUL_NO_OVERFLOW || size >= MUL_NO_OVERFLOW) &&
 	    nmemb > 0 && SIZE_MAX / nmemb < size) {
```

Comment:
One could change this conditional to:

        if (size > 0 && nmemb > 0 &&
            (nmemb >= MUL_NO_OVERFLOW || size >= MUL_NO_OVERFLOW) &&
            SIZE_MAX / nmemb < size)
                return nil;